### PR TITLE
Add evaluation on identifiers

### PR DESCRIPTION
**Pull request by irinakhismatullina**
1. Add evaluation on solo identifiers.
2. Add Candidates class usage in some more missed places.
3. Add `"processes number:` parameter to the analyzer config.

### CODE REVIEW

File: lookout/style/typos/benchmarks/evaluate_typos.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return template.render(generate_report=generate_report,
+                           vocabulary_tokens=analyzer.corrector.generator.tokens,
+                           n_candidates=analyzer.config["n_candidates"],
+                           IDENTIFIER_INDEX_COLUMN=IDENTIFIER_INDEX_COLUMN,
+                           Candidate=Candidate, Columns=Columns,
+                           flatten_df_by_column=flatten_df_by_column,
+                           test=test,
+                           suggestions=suggestions,
+                           tokenize=lambda x: " ".join(list(analyzer.parser.split(x))),
+                           str=str, set=set)
```
Comment (zurk) [inline lookout/style/typos/benchmarks/evaluate_typos.py]: I did not look through all PR yet but I think we should add CLI to call this function.
In this case, we also do not need to specify `TYPOS_DATASET`.

If you want you can also add a command run to the Makefile.

File: lookout/style/typos/cmdline_tools.py [hunk 2]
```diff
@@ -175,6 +176,20 @@ def add_parser(name, help):
               recommended to set this parameter if there are more than 20 repositories \
               in the dataset. Temporary directory is used if not set.")
 
+    # Report for typos on identifiers datasets
+    typos_report_parser = add_parser("typos-identifiers-report",
```
Comment (vmarkovtsev) [inline lookout/style/typos/cmdline_tools.py]:
```suggestion
    typos_report_parser = add_parser("evaluate-fixes",
```

"evaluate-noise" for testing on correct identifiers

File: lookout/style/typos/cmdline_tools.py [hunk 3]
```diff
@@ -175,6 +176,20 @@ def add_parser(name, help):
...
+    typos_report_parser.set_defaults(handler=evaluate_typos_on_identifiers)
+    add_config_arg(typos_report_parser)
+    typos_report_parser.add_argument(
+        "-d", "--dataset", required=False,
+        help="csv file with with typos. Must contain wrong_id and correct_id as the first"
+             "two columns. It is possible to specify the xz compressed file. By default the "
+             "identifiers from the Typo Commits Dataset are used.")
+    typos_report_parser.add_argument(
+        "-o", "--mistakes-output", required=False,
+        help="csv file where to put the wrong corrections.")
```
Comment (vmarkovtsev) [inline lookout/style/typos/cmdline_tools.py]:
```suggestion
        help="CSV file where to write the wrong corrections.")
```

File: lookout/style/typos/cmdline_tools.py [hunk 4]
```diff
@@ -175,6 +176,20 @@ def add_parser(name, help):
...
               in the dataset. Temporary directory is used if not set.")
 
+    # Report for typos on identifiers datasets
+    typos_report_parser = add_parser("typos-identifiers-report",
+                                     "Generate report for typo-ed identifiers dataset.")
+    typos_report_parser.set_defaults(handler=evaluate_typos_on_identifiers)
+    add_config_arg(typos_report_parser)
+    typos_report_parser.add_argument(
+        "-d", "--dataset", required=False,
+        help="csv file with with typos. Must contain wrong_id and correct_id as the first"
```
Comment (vmarkovtsev) [inline lookout/style/typos/cmdline_tools.py]:
```suggestion
        help="CSV file with with typos. The first two columns are wrong_id and correct_id."
```

File: lookout/style/typos/tests/test_evaluate_typos.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+TEST_DATA_PATH = pathlib.Path(__file__).parent
+
+
+class TyposCorrectorTest(unittest.TestCase):
+    def test_evaluate_typos_on_identifiers(self):
+        config = {"processes_number": 8, "corrector": str(TEST_DATA_PATH / "test_corrector.asdf")}
+        s = evaluate_typos_on_identifiers(str(TEST_DATA_PATH / "test_commits_with_typo.csv.xz"),
+                                          config)
+        self.assertGreater(len(s), 0)
+        print(s)
```
Comment (vmarkovtsev) [inline lookout/style/typos/tests/test_evaluate_typos.py]:
```suggestion
```

File: lookout/style/typos/benchmarks/evaluate_typos.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
```
Comment (vmarkovtsev) [inline lookout/style/typos/benchmarks/evaluate_typos.py]:
```suggestion
                           set=set)
```

File: lookout/style/typos/benchmarks/evaluate_typos.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        test[test["sugg 0"] != test.correct][["wrong", "sugg 0", "correct"]].to_csv(
+            mistakes_output)
+    template = load_jinja2_template(os.path.join(TEMPLATE_DIR, "quality_on_identifiers.md.jinja2"))
+    return template.render(generate_report=generate_report,
+                           vocabulary_tokens=analyzer.corrector.generator.tokens,
+                           n_candidates=analyzer.config["n_candidates"],
+                           IDENTIFIER_INDEX_COLUMN=IDENTIFIER_INDEX_COLUMN,
+                           Candidate=Candidate, Columns=Columns,
+                           flatten_df_by_column=flatten_df_by_column,
+                           test=test,
```
Comment (vmarkovtsev) [inline lookout/style/typos/benchmarks/evaluate_typos.py]: Let's rename `test` to smth else, very confusing name.
Also please sort: first objects, second functions

File: lookout/style/typos/templates/quality_on_identifiers.md.jinja2 [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+## Separate tokens report
+
+{{ generate_report(flat_test, flat_suggestions) }}
+
+## Identifiers correction quality
+
+{% set s = {"cumsum": 0} %}
+{% for pos in range(n_candidates) %}
+{% do s.__setitem__("cumsum", s["cumsum"] + 100 * (test["sugg " + str(pos)].eq(test["correct"]).mean())) %}
```
Comment (vmarkovtsev) [inline lookout/style/typos/templates/quality_on_identifiers.md.jinja2]:
```suggestion
{% do s.__setitem__("cumsum", s["cumsum"] + 100 * (test["sugg " + (pos | string)].eq(test["correct"]).mean())) %}
```

File: lookout/style/typos/templates/quality_on_identifiers.md.jinja2 [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+{% set flat_correct = flatten_df_by_column(test, Columns.CorrectSplit, Columns.CorrectToken, str.split) %}
+{% do flat_test.__setitem__(Columns.CorrectToken, flat_correct[Columns.CorrectToken]) %}
+{% set flat_suggestions = {} %}
+{% for i, row in flat_test.iterrows() %}
+        {% do flat_suggestions.__setitem__(i, suggestions.get(row[IDENTIFIER_INDEX_COLUMN], {}).get(
+        row[Columns.Token], [Candidate(row[Columns.Token], 1.0)]))%}
+{% endfor %}
+## Vocabulary insights
+
+{% set correct_tokens = set(flat_test.correct_token) %}
```
Comment (vmarkovtsev) [inline lookout/style/typos/templates/quality_on_identifiers.md.jinja2]:
```suggestion
{% set correct_tokens = frozenset(flat_test.correct_token) %}
```

As far as I understand `frozenset` is a built-in - need to check

Comment (irinakhismatullina) [reply lookout/style/typos/templates/quality_on_identifiers.md.jinja2]: Can't find it anywhere, I see only lists and dicts and tuples are supported.

**Comment by irinakhismatullina** [comment]
Ok, so gensim had a release 20 minutes ago, and now it's incopatible with our code. @vmarkovtsev should I rewrite it now to be up-to-date, or to specify lower upper-boundary on the package version?

**Comment by vmarkovtsev** [comment]
It depends on how much changes it requires. Rewriting everything => then definitely the version limit.

File: lookout/style/typos/templates/quality_on_identifiers.md.jinja2 (hunk 9 above)
Comment (vmarkovtsev) [reply lookout/style/typos/templates/quality_on_identifiers.md.jinja2]: Since better to have a frozen one - immutability is always a good idea.

**Comment by irinakhismatullina** [comment]
It doesn't look like it's something big, but it may take some time.

**Comment by irinakhismatullina** [comment]
Limited the version, I suggest moving update to the separate PR.

File: lookout/style/typos/templates/quality_on_identifiers.md.jinja2 [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+{% set flat_correct = flatten_df_by_column(identifiers, "correct", Columns.CorrectToken, tokenize) %}
+{% do flat_identifiers.__setitem__(Columns.CorrectToken, flat_correct[Columns.CorrectToken]) %}
+{% set flat_suggestions = {} %}
+{% for i, row in flat_identifiers.iterrows() %}
+        {% do flat_suggestions.__setitem__(i, suggestions.get(row[IDENTIFIER_INDEX_COLUMN], {}).get(
+        row[Columns.Token], [Candidate(row[Columns.Token], 1.0)]))%}
+{% endfor %}
+## Vocabulary insights
+
+{% set correct_tokens = set(flat_identifiers.correct_token) %}
```
Comment (vmarkovtsev) [inline lookout/style/typos/templates/quality_on_identifiers.md.jinja2]: OK, I found that there is the `unique` filter. We should use it instead.

**Pull request merged**